language/sygus.tex

```diff
@@ -2402,12 +2403,73 @@ \subsection{Oracle Example}
 %
 The solver calls the oracle with the command ${\tt N \,\,textfile.oracle}$, where ${\tt N}$ is the name of the oracle implememtation as specified in the oracle constraint declaration.
 
-\section{Objectives}%
-\label{sec:objectives}
+\section{Attributes}
+\label{sec:attr}
+
+In this section, we provide standardize attributes that are specific to
+the features introduced in this document.
+
+\subsection{Weights}%
+\label{sec:attr-weights}
+
+\begin{enumerate}
+\item The attribute $\texttt{:default}$
+expects an integer attribute value $n$.
+It specifies that the default weight value of declared weight keyword is $n$.
+\end{enumerate}
+\noindent
+Recall from \cref{sec:weight-semantics} that the default value
+for weight keywords is $0$.
+On the other hand, the command:
+\begin{lstlisting}[language=SyGuS, numbers=none, basicstyle=\ttfamily\small]
+(declare-weight numOps :default 1)
+\end{lstlisting}
+declares a weight keyword $\texttt{numOps}$ whose default weight value is $1$.
+
+\subsection{Objectives}%
+\label{sec:attr-objectives}
 
-In this section, we standardize several common use cases of the $\optsynthkwd$ command.
+In this section, we standardize attributes that pertain to specifying objectives of the $\optsynthkwd$ command.
+We classify two kinds of attributes.
+The first pertain to the orderings on values,
+which are supplied in the first argument of $\optsynthkwd$ command.
+The second pertain to orderings on the tuple itself.
+Below, we use $\succ$ to denote orderings on values or tuples of values,
+such that when $e_1 \succ e_2$, we have that $e_1$ is a more preferred solution.
 
-TODO
+\paragraph{Value Orderings}
+\begin{enumerate}
+\item The attribute $\texttt{:max}$ specifies that the ordering on values 
+$\succ$ for a given term $t$ should be maximized according to the default ordering
+of the type of $t$.
+The specification is well-defined only if the type of $t$ has a default ordering.
+We assume the default ordering $>$ is used for reals and integers, such that
+$v_1 \succ v_2$ if $v_1 > v_2$.
+Orderings on other types are not standardized by this document.
+Note that this attribute expects no provided attribute value.
+\item The attribute $\texttt{:min}$ is dual to $\texttt{:max}$, such that that ordering $\succ$
+should minimize the value based on the default ordering of its type.
+Like $\texttt{:max}$, this attribute expects no provided attribute value.
+\end{enumerate}
+
+\paragraph{Tuple Orderings}
+
+\begin{enumerate}
+\item The attribute $\texttt{:lexico}$ specifies that the ordering on tuples is lexicographic.
+In particular, assuming orderings $\succ_1, \ldots, \succ_n$ on the components of $n$-tuples,
+$(c_1, \ldots, c_n) \succ (d_1, \ldots, d_n)$ holds if there exists a $0 \leq j \leq n$
+such that $c_i = d_i$ for $i<j$ and $c_j \succ_j d_j$.
+This attribute expects no provided attribute value.
```

Comment:
One question here is, what should be the behavior when a value is provided after this attribute? Would this be a syntax error, or would it be ignored?

Reply:
This would be a syntax error.

Reply:
Thanks Andy. Should we add a sentence about this in the syntax section may be?

I've also approved this PR, so please feel free to merge whenever you're happy with it.

Reply:
My intention was to make this section the reference on the concrete syntax, so this sentence should be taken as the authority.

Reply:
I see, that makes sense too. Thanks.